src/verifier/step4/EqPolynomial.sol

```diff
@@ -0,0 +1,161 @@
+pragma solidity ^0.8.0;
+
+import "src/pasta/Vesta.sol";
+import "src/pasta/Pallas.sol";
+
+library EqPolinomialLib {
+    function evaluateVesta(uint256[] memory r, uint256[] memory rx) public pure returns (uint256){
+        require(r.length == rx.length, "[EqPolinomialLib.evaluateVesta] wrong input data length");
+
+
+        uint256 modulusVesta = Vesta.P_MOD;
+        uint256 modulusPallas = Pallas.P_MOD;
+
+        uint256 resultIter = 0;
+        uint256 rx_inner = 0;
+        uint256 r_inner = 0;
+        uint256 tmp1 = 0;
+        uint256 tmp2 = 0;
+        uint256 tmp3 = 0;
+        uint256 tmp4 = 0;
+        uint256 minus_rx = 0;
+        uint256 minus_r = 0;
+
+        uint256 result = 1;
+        for (uint256 i = 0; i < rx.length; i++) {
+            rx_inner = rx[i];
+            r_inner = r[i];
+            minus_rx = Pallas.negate(rx_inner);
+            minus_r = Pallas.negate(r_inner);
+            assembly {
+            // rx[i] * r[i] (Vesta!)
+                tmp1 := mulmod(rx_inner, r_inner, modulusVesta)
+            // 1 - rx[i] (Pallas!)
+                tmp2 := addmod(1, minus_rx, modulusPallas)
+            // 1 - r[i] (Pallas!)
+                tmp3 := addmod(1, minus_r, modulusPallas)
```

Comment:
Why should this arithmetic be done modulo the Pallas prime, and not the Vesta prime?

Reply:
I'm not enough qualified to explain it "formally", but I suspect it is related to the cycle property of Pallas and Vesta curves - e.g. these curves are highly 2-adic and the order of each curve is exactly the base field of the other. These resources will probably give you more sensible information:
- Pasta curves [blog](https://electriccoin.co/blog/the-pasta-curves-for-halo-2-and-beyond/) from ECC;
- Section 6.1 of Halo-2 [paper](https://eprint.iacr.org/2019/1021.pdf).

@huitseeker 's input may also be useful. From my engineering side, I can say that EqPolynomial is completely deterministic, e.g. two evaluations of same input will give exactly same outputs. Having that, one can always extract `r_x`, `tau` and `tau_bound_rx` values by running our hardcoded proving example and define a test vectors for [primary](https://github.com/lurk-lab/solidity-verifier/blob/eq-polynomial/test/eq-evaluation-tests.t.sol#L10) and [secondary](https://github.com/lurk-lab/solidity-verifier/blob/eq-polynomial/test/eq-evaluation-tests.t.sol#L49) evaluations. Having those test vectors you can always perform debugging by comparing the output that smart contract gives with the output given by reference `EqPolynomial::evaluate` implementation.

Reply:
OK, so the original line is here: https://github.com/microsoft/Nova/blob/031738de5127dbe1cc19ce5a42f7698c2dddeea0/src/spartan/polynomial.rs#L21
As the IDE UI makes evident (but the signature makes this clear as well), there's only one Scalar type involved:
![Screenshot 2023-06-13 at 11 09 42 AM](https://github.com/lurk-lab/solidity-verifier/assets/4142/650eb9a8-72af-4380-96cb-ed86b46b779d)

The source for the involved equation is in the proof of Lemma 2.1 in the Nova paper. There seem to be no justification for using a distinct modulus, outside of the scalar field involved to define the polynomials (i.e. the modulus of the Vesta Scalar field for polys over Vesta, the modulus of the Pallas Scalar field for polys over Pallas).

Now, note, as we discussed above:
- the modulus of the Vesta Scalar field is the modulus of the Pallas base field,
- the modulus of the Pallas Scalar field is the modulus of the Vesta base field,

So the correct way to use a variable such as `modulus(Pallas|Vesta)` IMHO would be to rename to what's actually meant, i.e. `scalarModulus(Pallas|Vesta)` and making sure the correct value is used.